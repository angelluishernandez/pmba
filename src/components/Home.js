import React, { useState, useEffect } from "react";
import { Input } from "./styled-components/Input";
import { ButtonContainer } from "./styled-components/ButtonContainer";
import { connect } from "react-redux";
import { addMovie } from "../redux/actions/movies.actions";
import MoviesList from "./MoviesList";
import { v4 as uuidv4 } from "uuid";
import FilterBar from "./FilterBar";

const Home = ({ addMovie, movies }) => {
	const [movieName, setMovieName] = useState("");
	const [movieGenres, setMovieGenres] = useState([]);
	const [movie, setMovie] = useState({});
	const genresFilter = ["Horror", "Romance", "Comedy"];
	const [filterGenres, setFilterGenres] = useState([]);

	useEffect(() => {
		if (Object.keys(movie).length !== 0) {
			addMovie(movie);
		}
	}, [movie]);

	// Handle new movie submit

	const handleSubmit = (e) => {
		e.preventDefault();

		const movieGenresArr = movieGenres.split(/[ ,]+/);
		const moviesToLowerCase = movieGenresArr.map((movie) =>
			movie.toLowerCase()
		);

		const reducedGenres = moviesToLowerCase.reduce(
			(movieArr, currentMovie) =>
				movieArr.includes(currentMovie)
					? movieArr
					: [...movieArr, currentMovie],
			[]
		);

		setMovie({
			name: movieName,
			movieGenres: reducedGenres,
			watched: false,
			id: uuidv4(),
		});
	};

	// Handle filter radio changes

	const handleRadioChange = (e) => {
		const value = e.target.value.toLowerCase();
		setFilterGenres([...filterGenres, value]);
	};

	return (
		<React.Fragment>
			<FilterBar
				filterGenres={filterGenres}
				setFilterGenres={setFilterGenres}
				handleRadioChange={handleRadioChange}
			/>
			{/* // This should have its own component */}

			<form onSubmit={handleSubmit}>
				<div>
					<label htmlFor="movieName">Movie name</label>
					<Input
						type="text"
						onChange={(e) => setMovieName(e.target.value)}
						value={movieName}
					/>
				</div>

				<div>
					<label htmlFor="movieGenres">Movie genres</label>
					<Input
						type="text"
						onChange={(e) => setMovieGenres(e.target.value)}
						value={movieGenres}
					/>
				</div>
				<ButtonContainer>Add a movie</ButtonContainer>
			</form>

			<MoviesList movies={movies} />
		</React.Fragment>
	);
};

const mapStateToProps = (state) => ({
	movies: state.movies,
});

const mapDispatchToProps = (dispatch) => ({
	addMovie: (movie) => dispatch(addMovie(movie)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Home);
