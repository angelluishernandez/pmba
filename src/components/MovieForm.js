import React from "react";
import { Input } from "./styled-components/Input";
import { ButtonContainer } from "./styled-components/ButtonContainer";
import { FormWrapper } from "./styled-components/FormWrapper";
import { FormItemWrapper } from "./styled-components/FormItemWrapper";

const MovieForm = ({
	handleSubmit,
	setMovieName,
	movieName,
	setMovieGenres,
	movieGenres,
}) => {
	return (
		<FormWrapper onSubmit={handleSubmit}>
			<h4>Create a movie</h4>
			<FormItemWrapper>
				<label htmlFor="movieName">Movie name</label>
				<Input
					type="text"
					onChange={(e) => setMovieName(e.target.value)}
					value={movieName}
				/>
			</FormItemWrapper>

			<FormItemWrapper>
				<label htmlFor="movieGenres">Movie genres</label>
				<Input
					type="text"
					onChange={(e) => setMovieGenres(e.target.value)}
					value={movieGenres}
				/>
				<ButtonContainer>Add a movie</ButtonContainer>
			</FormItemWrapper>
		</FormWrapper>
	);
};

export default MovieForm;
