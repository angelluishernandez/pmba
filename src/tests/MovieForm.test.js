import { configure, shallow, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import toJSON from "enzyme-to-json";
import React from "react";
import configureMockStore from "redux-mock-store";

import MovieForm from "../components/MovieForm";
import Home from "../components/Home";

const middlewares = [];
const createMockStore = configureMockStore(middlewares);

configure({ adapter: new Adapter() });

test("should render Movieform correctly", () => {
	const wrapper = shallow(<MovieForm />);
	expect(toJSON(wrapper)).toMatchSnapshot();
});

test("should change movie name in parent component on input change", () => {
	const initialState = { movies: [], filters: [] };
	const store = createMockStore(initialState);
	const value = "movieName";
	const wrapper = mount(<Home store={store} />);
	const input = wrapper.find(<MovieForm movieName={""} />).at(0);
	input.simulate("change", { target: { value } });
	expect(input.prop("movieName")).toEqual(value);
});
