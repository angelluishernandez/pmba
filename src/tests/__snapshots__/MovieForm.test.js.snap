// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Movieform correctly 1`] = `
<form>
  <div>
    <label
      htmlFor="movieName"
    >
      Movie name
    </label>
    <styled.input
      onChange={[Function]}
      type="text"
    />
  </div>
  <div>
    <label
      htmlFor="movieGenres"
    >
      Movie genres
    </label>
    <styled.input
      onChange={[Function]}
      type="text"
    />
  </div>
  <styled.button>
    Add a movie
  </styled.button>
</form>
`;
